console/src/state/resources.rs

```diff
@@ -0,0 +1,332 @@
+use crate::state::{truncate_registry_path, Field, Metadata};
+use crate::view;
+use console_api as proto;
+use std::{
+    cell::RefCell,
+    collections::HashMap,
+    convert::{TryFrom, TryInto},
+    rc::{Rc, Weak},
+    sync::Arc,
+    time::{Duration, SystemTime},
+};
+use tui::{
+    style::{Color, Modifier},
+    text::Span,
+};
+
+#[derive(Default, Debug)]
+pub(crate) struct ResourcesState {
+    resources: HashMap<u64, Rc<RefCell<Resource>>>,
+    new_resources: Vec<ResourceRef>,
+}
+
+#[derive(Debug, Copy, Clone)]
+#[repr(usize)]
+pub(crate) enum SortBy {
+    Rid = 0,
+    Kind = 1,
+    ConcreteType = 2,
+    Target = 3,
+    Total = 4,
+}
+
+#[derive(Debug, Clone, Eq, PartialEq, Ord, PartialOrd)]
+pub(crate) enum Kind {
+    Timer,
+    Other(String),
+}
+
+#[derive(Debug)]
+pub(crate) struct Resource {
+    id: u64,
+    meta_id: u64,
+    kind: Kind,
+    stats: ResourceStats,
+    target: Arc<str>,
+    concrete_type: String,
+    location: String,
+}
+
+pub(crate) type ResourceRef = Weak<RefCell<Resource>>;
+
+#[derive(Debug)]
+struct ResourceStats {
+    created_at: SystemTime,
+    dropped_at: Option<SystemTime>,
+    total: Option<Duration>,
+    attributes: Vec<Attribute>,
+    formatted_attributes: Vec<Vec<Span<'static>>>,
+}
+
+#[derive(Debug)]
+pub(crate) struct Attribute {
+    field: Field,
+    unit: Option<String>,
+}
+
+impl Default for SortBy {
+    fn default() -> Self {
+        Self::Rid
+    }
+}
+
+impl SortBy {
+    pub fn sort(&self, now: SystemTime, resources: &mut Vec<Weak<RefCell<Resource>>>) {
+        match self {
+            Self::Rid => {
+                resources.sort_unstable_by_key(|resource| resource.upgrade().map(|r| r.borrow().id))
+            }
+            Self::Kind => resources.sort_unstable_by_key(|resource| {
+                resource.upgrade().map(|r| r.borrow().kind.clone())
+            }),
+            Self::ConcreteType => resources.sort_unstable_by_key(|resource| {
+                resource.upgrade().map(|r| r.borrow().concrete_type.clone())
```

Comment:
it's a bummer that we have to clone all the concrete type strings _every_ time we sort the resources. we might want to consider `Arc`ing them to make this lighter-weight, since we can't borrow them as `&str`s because they would outlive the `RefCell`'s `Ref` lifetime...

i also have some future ideas about possible string interning but that's best saved for a follow-up.